### PR TITLE
[MLv2] remove & replace breakout clause

### PR DESCRIPTION
**Pull request by snoe**
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/29777)

### CODE REVIEW

File: src/metabase/lib/query.cljc [hunk 1]
```diff
@@ -8,11 +8,68 @@
...
+    (loop [stage-number stage-number]
+      (when-let [next-stage (lib.util/next-stage-number query stage-number)]
+        (when-not (->> (lib.stage/visible-columns query next-stage)
+                       (some (fn [{:keys [lib/source id] :as _column}]
+                               (and (= :source/previous-stage source) (= target-id id)))))
+          ;; Get the ref to look for from the previous-query
+          (let [target-ref (->> (lib.stage/visible-columns previous-query next-stage)
+                                (some (fn [{:keys [lib/source id] :as column}]
+                                        (when (and (= :source/previous-stage source) (= target-id id))
+                                          (lib.ref/ref column)))))]
```
Comment (snoe) [inline src/metabase/lib/query.cljc:46]: I'm not sure if there's a better way to do this, if I have a clause in stage n, then I need to know how it was referenced in stage n+1.

File: src/metabase/lib/util.cljc [hunk 2]
```diff
@@ -57,15 +58,16 @@
...
-      (dissoc stage location))))
+  (if-let [target (get stage location)]
+    (let [target-uuid (clause-uuid target-clause)
+          result (->> target
+                      (remove (comp #{target-uuid} clause-uuid))
+                      vec
+                      not-empty)]
+      (if result
+        (assoc stage location result)
+        (dissoc stage location)))
```
Comment (metamben) [inline src/metabase/lib/util.cljc:69]:
```suggestion
          result (into [] (remove (comp #{target-uuid} clause-uuid)) target)]
      (if (seq result)
        (assoc stage location result)
        (dissoc stage location)))
```

File: src/metabase/lib/query.cljc [hunk 3]
```diff
@@ -29,28 +86,15 @@
      ;; `location` should probably be "found" first before iterating.
      (reduce
        (fn [query location]
-         (lib.util/update-query-stage query stage-number
-                                      lib.util/replace-clause location target-clause replacement))
+         (let [result (lib.util/update-query-stage query stage-number
+                                                   lib.util/replace-clause location target-clause replacement)]
```
Comment (metamben) [inline src/metabase/lib/query.cljc:90]: This seems to be the only difference from `remove-clause`, so it might make sense to have a shared implementation.

File: src/metabase/lib/query.cljc [hunk 4]
```diff
@@ -8,11 +8,68 @@
...
+    target-clause]
+   (remove-clause query -1 target-clause))
+  ([query :- :metabase.lib.schema/query
+    stage-number :- :int
+    target-clause]
+   (reduce
+     (fn [query location]
+       (let [result (lib.util/update-query-stage query stage-number
+                                                 lib.util/remove-clause location target-clause)]
+         (when (not= query result)
```
Comment (metamben) [inline src/metabase/lib/query.cljc:65]: We could stop the reduction at this point, right?

File: src/metabase/lib/query.cljc [hunk 5]
```diff
@@ -8,11 +8,68 @@
...
    [metabase.lib.schema.id :as lib.schema.id]
+   [metabase.lib.stage :as lib.stage]
    [metabase.lib.util :as lib.util]
+   [metabase.mbql.util.match :as mbql.match]
    [metabase.util.malli :as mu]))
 
+(defn- find-clause
+  [query stage-number target-clause]
+  (let [stage (lib.util/query-stage query stage-number)
+        [target-type _opts target-id] target-clause]
```
Comment (metamben) [inline src/metabase/lib/query.cljc:22]: What kind of ID is `target-id`?

**Review by bshepherdson — CHANGES_REQUESTED** [review]
> Reviewed 4 of 6 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29777)* status: all files reviewed, 4 unresolved discussions (waiting on @snoe)

___
*[`src/metabase/lib/query.cljc` line 46 at r2](https://reviewable.io/reviews/metabase/metabase/29777#-NSMj_3R42EOGCq0OHeQ:-NSMj_3R42EOGCq0OHeR:b5xm3bo) ([raw file](https://github.com/metabase/metabase/blob/a75681ed1172223b773a16d13b13db33581fe380/src/metabase/lib/query.cljc#L46)):*
> ```Clojure
>                                 (some (fn [{:keys [lib/source id] :as column}]
>                                         (when (and (= :source/previous-stage source) (= target-id id))
>                                           (lib.ref/ref column)))))]
> ```

This condition appears a few times, I think it can be named and extracted, which would make this easier to read.

`(lib.stage/references-previous-column query stage-number target-id)` returning `column`, and wrap the result with `lib.ref/ref` for this second use.



_Suggestion:_
```Clojure
          (->> (lib.stage/visible-columns previous-query next-stage)
                                (some (fn [{:keys [lib/source id] :as column}]
                                        (when (and (= :source/previous-stage source) (= target-id id))
                                          (lib.ref/ref column))))
```
___
*[`src/metabase/lib/query.cljc` line 71 at r2](https://reviewable.io/reviews/metabase/metabase/29777#-NSMkd7B-Las9TRgrt0c:-NSMkd7B-Las9TRgrt0d:b97kdi3) ([raw file](https://github.com/metabase/metabase/blob/a75681ed1172223b773a16d13b13db33581fe380/src/metabase/lib/query.cljc#L71)):*
> ```Clojure
>          result))
>      query
>      [:order-by :breakout])))
> ```

I'm not sure about this explicit list. Why only these fields? It's okay if there's a reason, but we should write it down.

File: src/metabase/lib/query.cljc [hunk 6]
```diff
@@ -8,11 +8,68 @@
...
+      (reduce
+        (fn [accum top-level-clause]
+          (if (mbql.match/match-one (get stage top-level-clause)
+                [target-type _ target-id] true)
+            (conj accum top-level-clause)
+            accum))
+        #{}
+        [:order-by :aggregation :breakout :filter :expressions :joins :fields]))))
+
+(defn- check-subsequent-stages!
```
Comment (metamben) [inline src/metabase/lib/query.cljc:33]: This name is not giving enough information about what is checked.

File: src/metabase/lib/query.cljc [hunk 7]
```diff
@@ -8,11 +8,68 @@
...
+  "Throws if target-clause is used in a subsequent stage"
+  [previous-query query stage-number target-clause]
+  (let [[_ _ target-id] target-clause]
+    (loop [stage-number stage-number]
+      (when-let [next-stage (lib.util/next-stage-number query stage-number)]
+        (when-not (->> (lib.stage/visible-columns query next-stage)
+                       (some (fn [{:keys [lib/source id] :as _column}]
+                               (and (= :source/previous-stage source) (= target-id id)))))
```
Comment (metamben) [inline src/metabase/lib/query.cljc:41]: I find this logic hard to follow. Isn't it enough to check if the removed clause is referenced in subsequent stages? Can it be visible even though it has been removed?

File: test/metabase/lib/query_test.cljc [hunk 8]
```diff
@@ -72,35 +72,103 @@
...
+(deftest ^:parallel replace-clause-breakout-by-test
+  (let [query (-> (lib/query-for-table-name meta/metadata-provider "VENUES")
+                  (lib/breakout (lib/field (meta/id :venues :id)))
+                  (lib/breakout (lib/field (meta/id :venues :name))))
+        breakouts (lib/current-breakouts query)
+        replaced (-> query
+                     (lib/replace-clause (first breakouts) (lib/field (meta/id :venues :price))))
+        replaced-breakouts (lib/current-breakouts replaced)]
+    (is (= 2 (count breakouts)))
+    (is (not= breakouts replaced-breakouts))
```
Comment (metamben) [inline test/metabase/lib/query_test.cljc:157]: Can't we have a more precise assertion about the first (replaced) breakout?

File: test/metabase/lib/query_test.cljc [hunk 9]
```diff
@@ -72,35 +72,103 @@
...
-                         (lib/replace-clause (first order-bys) (lib/order-by-clause (lib/field (meta/id :venues :id)))))
-            replaced-order-bys (lib/order-bys replaced)]
-        (is (not= order-bys replaced-order-bys))
-        (is (= 2 (count replaced-order-bys)))
-        (is (= (second order-bys) (second replaced-order-bys)))))))
+    (is (= 2 (count order-bys)))
+    (let [replaced (-> query
+                       (lib/replace-clause (first order-bys) (lib/order-by-clause (lib/field (meta/id :venues :id)))))
+          replaced-order-bys (lib/order-bys replaced)]
+      (is (not= order-bys replaced-order-bys))
```
Comment (metamben) [inline test/metabase/lib/query_test.cljc:144]: Can't we have a more precise assertion about the first (replaced) order-by?
Would replacing the second order-by work as expected?

File: src/metabase/lib/query.cljc (hunk 3 above)
Comment (snoe) [reply src/metabase/lib/query.cljc:94]: I want to implement more removal targets first, but I'll keep an eye open.

File: src/metabase/lib/query.cljc (hunk 7 above)
Comment (snoe) [reply src/metabase/lib/query.cljc]: Added a comment, but I think in the case of the final `breakout` or `fields` being removed, it's generally safe and not surprising to keep using the same reference transparently.

File: src/metabase/lib/breakout.cljc [hunk 10]
```diff
@@ -31,6 +31,14 @@
      (lib.util/update-query-stage query stage-number update :breakout (fn [breakouts]
                                                                         (conj (vec breakouts) expr))))))
 
+(mu/defn current-breakouts :- [:maybe [:sequential ::lib.schema.expression/expression]]
```
[the remark this answers is not in the document]
Comment (snoe) [reply src/metabase/lib/breakout.cljc:34]: This was a part of a broader discussion to drop `external-op`. I'm following what we have in place for `order-bys` as that's the only e2e implementation atm.

**Review by snoe — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29777)* status: 2 of 6 files reviewed, 13 unresolved discussions (waiting on @bshepherdson and @metamben)

___
*[`src/metabase/lib/query.cljc` line 33 at r2](https://reviewable.io/reviews/metabase/metabase/29777#gh-1160126587:-NSfsA_0ETFo8CR4h9Ka:b-896fix) ([raw file](https://github.com/metabase/metabase/blob/a75681ed1172223b773a16d13b13db33581fe380/src/metabase/lib/query.cljc#L33)):*
<details><summary><i>Previously, metamben wrote…</i></summary><blockquote>

This name is not giving enough information about what is checked.
</blockquote></details>

Done.
___
*[`src/metabase/lib/query.cljc` line 46 at r2](https://reviewable.io/reviews/metabase/metabase/29777#-NSMj_3R42EOGCq0OHeQ:-NSfs82y9ekN6EB6Vq2W:b-896fix) ([raw file](https://github.com/metabase/metabase/blob/a75681ed1172223b773a16d13b13db33581fe380/src/metabase/lib/query.cljc#L46)):*
<details><summary><i>Previously, bshepherdson (Braden Shepherdson) wrote…</i></summary><blockquote>

This condition appears a few times, I think it can be named and extracted, which would make this easier to read.

`(lib.stage/references-previous-column query stage-number target-id)` returning `column`, and wrap the result with `lib.ref/ref` for this second use.
</blockquote></details>

Done.
___
*[`src/metabase/lib/query.cljc` line 65 at r2](https://reviewable.io/reviews/metabase/metabase/29777#gh-1160106786:-NSfsE2X0UZrcR23b7z7:b-896fix) ([raw file](https://github.com/metabase/metabase/blob/a75681ed1172223b773a16d13b13db33581fe380/src/metabase/lib/query.cljc#L65)):*
<details><summary><i>Previously, metamben wrote…</i></summary><blockquote>

We could stop the reduction at this point, right?
</blockquote></details>

Done.
___
*[`src/metabase/lib/query.cljc` line 71 at r2](https://reviewable.io/reviews/metabase/metabase/29777#-NSMkd7B-Las9TRgrt0c:-NSfsEybAaeGCzmeGk3Y:b-8s3ecz) ([raw file](https://github.com/metabase/metabase/blob/a75681ed1172223b773a16d13b13db33581fe380/src/metabase/lib/query.cljc#L71)):*
<details><summary><i>Previously, bshepherdson (Braden Shepherdson) wrote…</i></summary><blockquote>

I'm not sure about this explicit list. Why only these fields? It's okay if there's a reason, but we should write it down.
</blockquote></details>

Done. I'm following up with other PRs to fill out all top level clauses since they generally need some pre-work.
___
*[`src/metabase/lib/util.cljc` line 69 at r2](https://reviewable.io/reviews/metabase/metabase/29777#gh-1160104594:-NSfsYZJEG3vYhpXLiii:b-896fix) ([raw file](https://github.com/metabase/metabase/blob/a75681ed1172223b773a16d13b13db33581fe380/src/metabase/lib/util.cljc#L69)):*
<details><summary><i>Previously, metamben wrote…</i></summary><blockquote>

```suggestion
          result (into [] (remove (comp #{target-uuid} clause-uuid)) target)]
      (if (seq result)
        (assoc stage location result)
        (dissoc stage location)))
```
</blockquote></details>

Done.
___
*[`src/metabase/lib/util.cljc` line 70 at r2](https://reviewable.io/reviews/metabase/metabase/29777#gh-1160101264:-NSfsYzV6wq_6lR5ngYi:b-896fix) ([raw file](https://github.com/metabase/metabase/blob/a75681ed1172223b773a16d13b13db33581fe380/src/metabase/lib/util.cljc#L70)):*
<details><summary><i>Previously, metamben wrote…</i></summary><blockquote>

The docstring should start with `Remove the`
</blockquote></details>

Done.
___
*[`test/metabase/lib/query_test.cljc` line 144 at r2](https://reviewable.io/reviews/metabase/metabase/29777#gh-1160133322:-NSfsZgh9z334_FIAyl9:b-896fix) ([raw file](https://github.com/metabase/metabase/blob/a75681ed1172223b773a16d13b13db33581fe380/test/metabase/lib/query_test.cljc#L144)):*
<details><summary><i>Previously, metamben wrote…</i></summary><blockquote>

Can't we have a more precise assertion about the first (replaced) order-by?
Would replacing the second order-by work as expected?
</blockquote></details>

Done.
___
*[`test/metabase/lib/query_test.cljc` line 157 at r2](https://reviewable.io/reviews/metabase/metabase/29777#gh-1160132457:-NSfs_9FBrVzVuyCIO-P:b-896fix) ([raw file](https://github.com/metabase/metabase/blob/a75681ed1172223b773a16d13b13db33581fe380/test/metabase/lib/query_test.cljc#L157)):*
<details><summary><i>Previously, metamben wrote…</i></summary><blockquote>

Can't we have a more precise assertion about the first (replaced) breakout?
</blockquote></details>

Done.

**Review by metamben — APPROVED** [review]
> Reviewed 1 of 3 files at r3, 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29777)* status: all files reviewed, 7 unresolved discussions (waiting on @bshepherdson and @snoe)

___
*[`src/metabase/lib/breakout.cljc` line 34 at r2](https://reviewable.io/reviews/metabase/metabase/29777#gh-1160086974:-NSfxRMjDV1DyEWPtfwh:bv9io8e) ([raw file](https://github.com/metabase/metabase/blob/a75681ed1172223b773a16d13b13db33581fe380/src/metabase/lib/breakout.cljc#L34)):*
<details><summary><i>Previously, snoe (Case Nelson) wrote…</i></summary><blockquote>

This was a part of a broader discussion to drop `external-op`. I'm following what we have in place for `order-bys` as that's the only e2e implementation atm.
</blockquote></details>

I've just opened #29928 which is related, I think. I'm not sure how important it is to keep CLJ(S) and JS in sync.
___
*[`src/metabase/lib/query.cljc` line 41 at r2](https://reviewable.io/reviews/metabase/metabase/29777#gh-1160129239:-NSfzNRK4NGV6_pV8R75:b-q7obpa) ([raw file](https://github.com/metabase/metabase/blob/a75681ed1172223b773a16d13b13db33581fe380/src/metabase/lib/query.cljc#L41)):*
<details><summary><i>Previously, snoe (Case Nelson) wrote…</i></summary><blockquote>

Added a comment, but I think in the case of the final `breakout` or `fields` being removed, it's generally safe and not surprising to keep using the same reference transparently.
</blockquote></details>

I see that removing clauses from `breakout` and `order-by` is fine, I'm not sure about `fields`.
___
*[`src/metabase/lib/query.cljc` line 90 at r2](https://reviewable.io/reviews/metabase/metabase/29777#gh-1160105903:-NSg1ur4BgoL19jEW7cb:b2nhktp) ([raw file](https://github.com/metabase/metabase/blob/a75681ed1172223b773a16d13b13db33581fe380/src/metabase/lib/query.cljc#L90)):*
<details><summary><i>Previously, snoe (Case Nelson) wrote…</i></summary><blockquote>

I want to implement more removal targets first, but I'll keep an eye open.
</blockquote></details>

OK.
___
*[`src/metabase/lib/query.cljc` line 37 at r4](https://reviewable.io/reviews/metabase/metabase/29777#-NSg13c2E-0ywdh8iQBt:-NSg0dZG3QKVkU4SUMug:bohqox) ([raw file](https://github.com/metabase/metabase/blob/15702136f2909006d81521f81452557ba1f79e0d/src/metabase/lib/query.cljc#L37)):*
> ```Clojure
>   "Gets the ref for the target-id exposed by the previous stage"
>   [query next-stage-number target-id]
>   (->> (lib.stage/visible-columns query next-stage-number)
> ```

Shouldn't the `next-` prefix be removed? Seeing both "next" and "previous" is confusing. It's not that we are working with a stage in between, is it?

**Review by bshepherdson — CHANGES_REQUESTED** [review]
> Reviewed 7 of 7 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29777)* status: all files reviewed, 7 unresolved discussions (waiting on @snoe)

___
*[`src/metabase/lib/remove_replace.cljc` line 22 at r6](https://reviewable.io/reviews/metabase/metabase/29777#-NSkYJOo0UQ_sr53RA9y:-NSkYJOo0UQ_sr53RA9z:undefined) ([raw file](https://github.com/metabase/metabase/blob/07ef46dd923220fdd5dd300322dfd1ccc482f626/src/metabase/lib/remove_replace.cljc#L22)):*
> ```Clojure
>             accum))
>         #{}
>         [:order-by :aggregation :breakout :filter :expressions :joins :fields]))))
> ```





_Suggestion:_
```Clojure
      (set (keep (fn [top-level-clause]
                   (mbql.match/match-one (get stage top-level-clause)
                     [target-type _ target-id] top-level-clause))
                 [:order-by :aggregation :breakout :filter :expressions :joins :fields])))))
```
___
*[`src/metabase/lib/remove_replace.cljc` line 37 at r6](https://reviewable.io/reviews/metabase/metabase/29777#-NSkZQDlAbGNVb-4W9QA:-NSkZQDlAbGNVb-4W9QB:btfbkeb) ([raw file](https://github.com/metabase/metabase/blob/07ef46dd923220fdd5dd300322dfd1ccc482f626/src/metabase/lib/remove_replace.cljc#L37)):*
> ```Clojure
>   (let [[_ _ target-id] target-clause]
>     (loop [stage-number stage-number]
>       (when-let [next-stage-number (lib.util/next-stage-number query stage-number)]
> ```

This `loop` feels like it would read better as a `(doseq [stage-number (tail-stages query stage-number)] ...)`. We're likely to want a sequence of all later stages several more times.

Actually it could probably be something like
```clojure
(when-let [target-ref (some (fn ...) (tail-stages query stage-number))]
  (throw ...))
```
___
*[`src/metabase/lib/remove_replace.cljc` line 60 at r6](https://reviewable.io/reviews/metabase/metabase/29777#-NSkaBsxDy4ojs0k326p:-NSkaBsxDy4ojs0k326q:b-7tw6vd) ([raw file](https://github.com/metabase/metabase/blob/07ef46dd923220fdd5dd300322dfd1ccc482f626/src/metabase/lib/remove_replace.cljc#L60)):*
> ```Clojure
>        (let [result (lib.util/update-query-stage query stage-number
>                                                  lib.util/remove-clause location target-clause)]
>          (when (not= query result)
> ```

`if`?
___
*[`src/metabase/lib/remove_replace.cljc` line 93 at r6](https://reviewable.io/reviews/metabase/metabase/29777#-NSkadhQ5T_DtONc3YTi:-NSkadhQ5T_DtONc3YTj:b5lmnay) ([raw file](https://github.com/metabase/metabase/blob/07ef46dd923220fdd5dd300322dfd1ccc482f626/src/metabase/lib/remove_replace.cljc#L93)):*
> ```Clojure
>        query
>        ;; TODO only these top level clauses are supported at this moment
>        [:order-by :breakout]))))
> ```

I really don't like this pattern, but I can't think of a better one. It's kind of a `clojure.walk` thing, crawling the tree looking for something, running a function on it when you find it, and returning to the top, but even that isn't a perfect fit. Maybe zipppers?



_Code quote:_
```Clojure
     (reduce
       (fn [query location]
         (let [result (lib.util/update-query-stage query stage-number
                                                   lib.util/replace-clause location target-clause replacement)]
           (when (not= query result)
             (case location
               :breakout (check-subsequent-stages-for-invalid-target! query result stage-number (lib.ref/ref target-clause))
               nil)
             (reduced result))
           result))
       query
       ;; TODO only these top level clauses are supported at this moment
       [:order-by :breakout]))))
```

**Comment by snoe** [comment]
*[`src/metabase/lib/remove_replace.cljc` line 37 at r6](https://reviewable.io/reviews/metabase/metabase/29777#-NSkZQDlAbGNVb-4W9QA:-NSknRnIAQfkTj8GKGBY:b3l45y) ([raw file](https://github.com/metabase/metabase/blob/07ef46dd923220fdd5dd300322dfd1ccc482f626/src/metabase/lib/remove_replace.cljc#L37)):*
<details><summary><i>Previously, bshepherdson (Braden Shepherdson) wrote…</i></summary><blockquote>

This `loop` feels like it would read better as a `(doseq [stage-number (tail-stages query stage-number)] ...)`. We're likely to want a sequence of all later stages several more times.

Actually it could probably be something like
```clojure
(when-let [target-ref (some (fn ...) (tail-stages query stage-number))]
  (throw ...))
```
</blockquote></details>

Possibly, there are some open issues around field refs that make it difficult to know if stage n+1 has to be treated differently from stage n+2.
I'll be in this code more for the other top level clauses and I'll keep this in mind.

**Comment by snoe** [comment]
*[`src/metabase/lib/query.cljc` line 41 at r2](https://reviewable.io/reviews/metabase/metabase/29777#gh-1160129239:-NSkpSVo0hm2mCs2s4JU:bx8sqkj) ([raw file](https://github.com/metabase/metabase/blob/a75681ed1172223b773a16d13b13db33581fe380/src/metabase/lib/query.cljc#L41)):*
<details><summary><i>Previously, metamben wrote…</i></summary><blockquote>

I see that removing clauses from `breakout` and `order-by` is fine, I'm not sure about `fields`.
</blockquote></details>

Default columns are exposed to the next stage when fields is removed
`:stages [{:source-table "venues" :fields [:field {} <id>]} {:expression {"foo" [:field {} "ID"]}}]` is still valid in this transform (IIUC)
`:stages [{:source-table "venues"} {:expression {"foo" [:field {} "ID"]}}]`

**Review by snoe — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29777)* status: 8 of 9 files reviewed, 7 unresolved discussions (waiting on @bshepherdson and @metamben)

___
*[`src/metabase/lib/remove_replace.cljc` line 22 at r6](https://reviewable.io/reviews/metabase/metabase/29777#-NSkYJOo0UQ_sr53RA9y:-NSkpI07BT_EYUXM4aE3:b-896fix) ([raw file](https://github.com/metabase/metabase/blob/07ef46dd923220fdd5dd300322dfd1ccc482f626/src/metabase/lib/remove_replace.cljc#L22)):*
> ```Clojure
>             accum))
>         #{}
>         [:order-by :aggregation :breakout :filter :expressions :joins :fields]))))
> ```

Done.

**Review by bshepherdson — APPROVED** [review]
> Reviewed 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29777)* status: all files reviewed, 4 unresolved discussions (waiting on @metamben and @snoe)

___
*[`src/metabase/lib/remove_replace.cljc` line 22 at r6](https://reviewable.io/reviews/metabase/metabase/29777#-NSkYJOo0UQ_sr53RA9y:-NSktnPZ2IIh0khlUrvh:b5r6ky8) ([raw file](https://github.com/metabase/metabase/blob/07ef46dd923220fdd5dd300322dfd1ccc482f626/src/metabase/lib/remove_replace.cljc#L22)):*
<details><summary><i>Previously, snoe (Case Nelson) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

nit: https://github.com/bbatsov/clojure-style-guide#threading-macros-and-optional-parentheses
___
*[`src/metabase/lib/remove_replace.cljc` line 60 at r6](https://reviewable.io/reviews/metabase/metabase/29777#-NSkaBsxDy4ojs0k326p:-NSku0GECZFz6vPvOgwm:bnbbrgg) ([raw file](https://github.com/metabase/metabase/blob/07ef46dd923220fdd5dd300322dfd1ccc482f626/src/metabase/lib/remove_replace.cljc#L60)):*
<details><summary><i>Previously, snoe (Case Nelson) wrote…</i></summary><blockquote>

I think the `when` is better at conveying it's only for checking the side effect.
</blockquote></details>

Ah, right, these are checks that throw. LGTM then.
